Add verbose pipeline parameter to output each processor's execution details

junweid62 opened the pull request:
Jan 16 revision: 

1. Add two new fields: status and error

- status: Represents the execution status of the processor, with possible values such as SUCCESS and FAIL.
- error: Captures the error message or details if the processor encounters an issue during execution. If no error occurs, this field remains empty.
2. When verbose pipeline is enabled, errors in a processor will not interrupt the search but will document the error and continue searching

When the verbose pipeline mode is enabled, if a processor encounters an error during execution, the search process will not be interrupted. Instead, the error will be documented in the processor's execution details (e.g., in ProcessorExecutionDetail) and the remaining search process will proceed as normal.

```
{
            "processor_name": "rename_field",
            "duration_millis": 0,
            "status": "fail",
            "error": "Document with id 1 is missing field messag123e",
            "input_data": [
                {
                    "_index": "my_index",
                    "_id": "1",
                    "_score": 1.0,
                    "_source": {
                        "message": "This is a public message",
                        "visibility": "public"
                    }
                },
                {
                    "_index": "my_index",
                    "_id": "2",
                    "_score": 1.0,
                    "_source": {
                        "message": "This is a private message",
                        "visibility": "private"
                    }
                }
            ],
            "output_data": null
        },
        {
```
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
3. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
4. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Related RFC : #16705 

This PR introduces enhancements to OpenSearch's search pipeline functionality, focusing on improving the traceability and debugging of search request and response transformations. It addresses the increasing complexity of search pipeline processors by implementing verbose mode support, which provides detailed insights into processor execution.

1. **Adds Verbose Mode for Search Pipelines**:
   - Introduced the `verbose_pipeline` parameter to search requests, default to false.
   - Tracks each processor’s input, output, execution time, and status (success/failure).
   - Provides detailed logs of the data flow through request and response processors.

2. **Improves Pipeline Debugging**:
   - Captures step-by-step data transformations applied by each processor.
   - Includes execution metadata (e.g., timestamps and elapsed time) in search responses for better analysis.

3. **Supports All Pipeline Configurations**:
   - Works seamlessly with:
     - Default pipelines configured at the index level.
     - Pipelines explicitly specified in the search request.
     - Ad-hoc pipelines defined inline.

5. **Test Framework Enhancements**:
   - Added tests for verbose mode to ensure correct functionality and compatibility.



Example output with request processor: `filter_query` response processor: `rename_field` and `sort`
```
{
    "took": 29,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 1,
            "relation": "eq"
        },
        "max_score": 0.9116078,
        "hits": [
            {
                "_index": "my_index",
                "_id": "2",
                "_score": 0.9116078,
                "_source": {
                    "notification": "This is a public message 2",
                    "visibility": "public"
                }
            }
        ]
    },
    "processor_result": [
        {
            "processor_name": "filter_query",
            "duration_millis": 0,
            "input_data": {
                "query": {
                    "bool": {
                        "must": [
                            {
                                "match": {
                                    "message": {
                                        "query": "This is a public message 1",
                                        "operator": "OR",
                                        "prefix_length": 0,
                                        "max_expansions": 50,
                                        "fuzzy_transpositions": true,
                                        "lenient": false,
                                        "zero_terms_query": "NONE",
                                        "auto_generate_synonyms_phrase_query": true,
                                        "boost": 1.0
                                    }
                                }
                            }
                        ],
                        "adjust_pure_negative": true,
                        "boost": 1.0
                    }
                },
                "search_pipeline": "my_pipeline"
            },
            "output_data": {
                "query": {
                    "bool": {
                        "must": [
                            {
                                "bool": {
                                    "must": [
                                        {
                                            "match": {
                                                "message": {
                                                    "query": "This is a public message 1",
                                                    "operator": "OR",
                                                    "prefix_length": 0,
                                                    "max_expansions": 50,
                                                    "fuzzy_transpositions": true,
                                                    "lenient": false,
                                                    "zero_terms_query": "NONE",
                                                    "auto_generate_synonyms_phrase_query": true,
                                                    "boost": 1.0
                                                }
                                            }
                                        }
                                    ],
                                    "adjust_pure_negative": true,
                                    "boost": 1.0
                                }
                            }
                        ],
                        "filter": [
                            {
                                "term": {
                                    "visibility": {
                                        "value": "public",
                                        "boost": 1.0
                                    }
                                }
                            }
                        ],
                        "adjust_pure_negative": true,
                        "boost": 1.0
                    }
                },
                "search_pipeline": "my_pipeline"
            }
        },
        {
            "processor_name": "rename_field",
            "duration_millis": 0,
            "input_data": [
                {
                    "_index": "my_index",
                    "_id": "2",
                    "_score": 0.9116078,
                    "_source": {
                        "message": "This is a public message 2",
                        "visibility": "public"
                    }
                }
            ],
            "output_data": [
                {
                    "_index": "my_index",
                    "_id": "2",
                    "_score": 0.9116078,
                    "_source": {
                        "notification": "This is a public message 2",
                        "visibility": "public"
                    }
                }
            ]
        },
        {
            "processor_name": "sort",
            "duration_millis": 0,
            "input_data": [
                {
                    "_index": "my_index",
                    "_id": "2",
                    "_score": 0.9116078,
                    "_source": {
                        "notification": "This is a public message 2",
                        "visibility": "public"
                    }
                }
            ],
            "output_data": [
                {
                    "_index": "my_index",
                    "_id": "2",
                    "_score": 0.9116078,
                    "_source": {
                        "notification": "This is a public message 2",
                        "visibility": "public"
                    }
                }
            ]
        }
    ]
}
```
### Related Issues
Resolves #14745 
<!-- List any other related issues here -->

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

junweid62 commented:
@joshpalis  Hey Josh, can you take a look when you have time? :)

junweid62 commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -225,11 +241,19 @@ ActionListener<SearchResponse> transformResponseListener(
 
             responseListener = ActionListener.wrap(r -> {
+                ProcessorExecutionDetail detail = new ProcessorExecutionDetail(processor.getType());
+                detail.addInput(Arrays.asList(r.getHits().deepCopy().getHits()));
```
I initially thought the same:). However, when the Object receives this list, it holds a reference to the original SearchHit objects. Without a deepCopy, any modifications to the output will also reflect in the input, as they share the same reference.

By using deepCopy, we ensure that the input and output hold independent values, allowing us to clearly see the differences between the two at each step of the processor.

junweid62 commented on server/src/main/java/org/opensearch/search/SearchHits.java:
```diff
@@ -166,6 +167,22 @@ public SearchHit[] getHits() {
+     * @return a deep copy of the current SearchHits object
+     * @throws IOException if an I/O exception occurs during serialization or deserialization
+     */
+    public SearchHits deepCopy() throws IOException {
```
This code implements deep copy by serializing (writeTo) and deserializing (StreamInput).

junweid62 commented:
Hi, @dbwiddis Could you please take a moment to review this PR when you have the time? I’d appreciate any feedback you might have :)

martin-gaievski commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Add new configuration setting `synonym_analyzer`, to the `synonym` and `synonym_graph` filters, enabling the specification of a custom analyzer for reading the synonym file ([#16488](https://github.com/opensearch-project/OpenSearch/pull/16488)).
 - Add stats for remote publication failure and move download failure stats to remote methods([#16682](https://github.com/opensearch-project/OpenSearch/pull/16682/))
 - Added a precaution to handle extreme date values during sorting to prevent `arithmetic_exception: long overflow` ([#16812](https://github.com/opensearch-project/OpenSearch/pull/16812)).
+- Add `verbose_pipeline` parameter to output each processor's execution details ([#14745](https://github.com/opensearch-project/OpenSearch/pull/14745)).
```
link has to point to PR https://github.com/opensearch-project/OpenSearch/pull/16843, current one is for the issue

martin-gaievski commented on server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -302,6 +305,9 @@ public SearchSourceBuilder(StreamInput in) throws IOException {
         if (in.getVersion().onOrAfter(Version.V_2_18_0)) {
             searchPipeline = in.readOptionalString();
         }
+        if (in.getVersion().onOrAfter(Version.CURRENT)) {
```
I think you need to have exact version here, CURRENT will change with every next release

junweid62 replied:
I initially used the exact version here, but it caused failures in the BWC integration tests. @dbwiddis  suggested switching to CURRENT for now, with the plan to update the version in a separate PR when preparing for the release.

martin-gaievski replied:
that makes sense, how about creating GH issue to track that future work, you can have the issue link as part of your TODO

owaiskazi19 replied:
This is a general BWC dance we do for such cases, we don't need an issue to track this. @junweid62 can take care of it by following the below

1. Do this on main with `onOrAfter(Version.V_3_0_0))`. Get it merged.
2. You'll need a manual backport to 2.x, where you do `onOrAfter(Version.V_2_19_0)`. Don't get it merged right away.
3. Before merging the backport to 2.x, open another PR on main to change it to onOrAfter(Version.V_2_19_0).
4. Merge the backport PR.
5. Merge the main version update PR.

martin-gaievski commented on server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -385,6 +391,9 @@ public void writeTo(StreamOutput out) throws IOException {
         if (out.getVersion().onOrAfter(Version.V_2_18_0)) {
             out.writeOptionalString(searchPipeline);
         }
+        if (out.getVersion().onOrAfter(Version.CURRENT)) {
```
same as in previous comment

martin-gaievski commented on server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1142,6 +1151,26 @@ public SearchSourceBuilder pipeline(String searchPipeline) {
+        return this;
+    }
+
+    public Boolean verbosePipeline() {
```
why return type is a wrapper around primitive Boolean?

martin-gaievski commented on server/src/main/java/org/opensearch/search/internal/InternalSearchResponse.java:
```diff
@@ -123,4 +147,15 @@ private static void writeSearchExtBuildersOnOrAfter(StreamOutput out, List<Searc
+
+    private static void writeProcessorResultOnOrAfter(StreamOutput out, List<ProcessorExecutionDetail> processorResult) throws IOException {
+        if (out.getVersion().onOrAfter(Version.V_2_18_0)) {
+            out.writeCollection(processorResult, (o, detail) -> detail.writeTo(o));
```
+1 to Josh's comment, please use `out.writeList()`

martin-gaievski commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -146,11 +147,18 @@ void transformRequest(SearchRequest request, ActionListener<SearchRequest> reque
             SearchRequestProcessor processor = searchRequestProcessors.get(i);
             currentListener = ActionListener.wrap(r -> {
+                ProcessorExecutionDetail detail = new ProcessorExecutionDetail(processor.getType());
+                detail.addInput(r.source().shallowCopy());
```
can you add check for `verbosePipeline` here, we're adding unnecessary overhead of doing shallow copy in case verbose is not enabled

martin-gaievski commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -225,11 +241,19 @@ ActionListener<SearchResponse> transformResponseListener(
 
             responseListener = ActionListener.wrap(r -> {
+                ProcessorExecutionDetail detail = new ProcessorExecutionDetail(processor.getType());
+                detail.addInput(Arrays.asList(r.getHits().deepCopy().getHits()));
```
we need to spend cycles on doing shallow copy only in case verbose flag is set, please consider making corresponding changes

martin-gaievski commented on server/src/main/java/org/opensearch/search/pipeline/PipelineProcessingContext.java:
```diff
@@ -35,4 +41,31 @@ public void setAttribute(String name, Object value) {
+        attributes.computeIfAbsent(PROCESSOR_EXECUTION_DETAILS_KEY, k -> new ArrayList<ProcessorExecutionDetail>());
+        List<ProcessorExecutionDetail> details = (List<ProcessorExecutionDetail>) attributes.get(PROCESSOR_EXECUTION_DETAILS_KEY);
+        details.add(detail);
+    }
```
Few extra cycles here for accessing map, how about this version:
```
public void addProcessorExecutionDetail(ProcessorExecutionDetail detail) {
    @SuppressWarnings("unchecked")
    List<ProcessorExecutionDetail> details = (List<ProcessorExecutionDetail>) attributes.computeIfAbsent(
        PROCESSOR_EXECUTION_DETAILS_KEY,
        k -> new ArrayList<>()
    );
    details.add(detail);
}
```

junweid62 replied:
Will change in the next revision

dbwiddis commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -225,11 +243,21 @@ ActionListener<SearchResponse> transformResponseListener(
                     long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - start);
                     afterResponseProcessor(processor, took);
+                    if (request.source().verbosePipeline()) {
+                        detail.addOutput(Arrays.asList(rr.getHits().deepCopy().getHits()));
```
Can you help me understand why this deepCopy is needed?

junweid62 replied:
When the Object receives this list, it holds a reference to the original SearchHit objects. Without a deepCopy, any modifications to the output will also reflect in the input, as they share the same reference.

By using deepCopy, we ensure that the input and output hold independent values, allowing us to clearly see the differences between the two at each step of the processor.

dbwiddis commented on server/src/main/java/org/opensearch/search/pipeline/PipelineProcessingContext.java:
```diff
@@ -17,6 +20,9 @@
     private final Map<String, Object> attributes = new HashMap<>();
 
+    // Key for processor execution details
+    private static final String PROCESSOR_EXECUTION_DETAILS_KEY = "processorExecutionDetails";
```
Something seems off to me about having this private constant in this class.  I see it's used in a getter so nobody needs to know what it is, but it's also a key in an attributes map that could have a name collision.

Are other attributes defined (should this constant be somewhere else, and public)?  If not, is re-using this map for an undocumented purpose a good idea?

junweid62 replied:
I noticed a similar usage in the [ScriptRequestProcessor class](https://github.com/opensearch-project/OpenSearch/blob/7ae66d0d63308b009927647ce0a892231f1fad29/modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/ScriptRequestProcessor.java#L149C9-L149C127), where keys are defined as:
```
private static final List<String> SCRIPT_CONFIG_KEYS = List.of("id", "source", "inline", "lang", "params", "options");
```
In our case, I’m considering changing the constant to public and moving it to the ProcessorExecutionDetail class 

Do you think this would be a better approach?

dbwiddis commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,246 @@
+ *
+ * @opensearch.internal
+ */
+@PublicApi(since = "1.0.0")
```
Is this new? Should this version be different?

junweid62 replied:
I took a closer look and feel that this annotation isn't necessary at all.

junweid62 replied:
already changed it to 2.19

dbwiddis commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,246 @@
+        return new ProcessorExecutionDetail(processorName, durationMillis, inputData, outputData);
+    }
+
+    private static Object parseFieldFromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token = parser.currentToken();
+        if (token == XContentParser.Token.VALUE_NULL) {
+            return null;
+        } else if (token == XContentParser.Token.START_ARRAY) {
+            return parseArrayFromXContent(parser);
+        } else if (token == XContentParser.Token.START_OBJECT) {
+            return parser.map();
+        } else {
+            return parser.textOrNull();
+        }
+    }
```
I would be surprised if this utility method (and the related array one) doesn't already exist somewhere ([like here?](https://github.com/opensearch-project/OpenSearch/blob/main/libs/x-content/src/main/java/org/opensearch/common/xcontent/XContentUtils.java)).  If it doesn't, perhaps this code should be in such a util class?

dbwiddis reviewed (commented):
LGTM generally.  Some comments and suggestions

martin-gaievski reviewed (approved):
Looks good to me, thank you

owaiskazi19 commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,221 @@
+    public static final ParseField PROCESSOR_NAME_FIELD = new ParseField("processor_name");
+    public static final ParseField DURATION_MILLIS_FIELD = new ParseField("duration_millis");
+    public static final ParseField INPUT_DATA_FIELD = new ParseField("input_data");
+    public static final ParseField OUTPUT_DATA_FIELD = new ParseField("output_data");
```
All the above vars can be changed to `Private` since we are not using them anywhere apart from this class

owaiskazi19 commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,221 @@
+            return;
+        }
+
+        if (data instanceof List) {
```
We could also have a case of `data instanceof Map`. Moreover, did we try all the [core processors](https://opensearch.org/docs/latest/search-plugins/search-pipelines/search-processors/) to come up with such cases?

owaiskazi19 commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,221 @@
+        long durationMillis = 0;
+        Object inputData = null;
+        Object outputData = null;
+        if (parser.currentToken() != XContentParser.Token.START_OBJECT) {
```
Why this extra check? We could directly check with `ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);`

junweid62 replied:
Thanks for pointing this out! The reason for including the null check here is consistency with how fromXContent is handled in other parts of the codebase. Some fromXContent implementation includes this check to handle cases where the parser starts in an unexpected state, ensuring robustness. 
ex:
https://github.com/opensearch-project/OpenSearch/blob/13159c1693d088417683f8d7f2a018ea7e8a6866/client/rest-high-level/src/main/java/org/opensearch/client/slm/GetSnapshotLifecyclePolicyResponse.java#L62
https://github.com/opensearch-project/OpenSearch/blob/13159c1693d088417683f8d7f2a018ea7e8a6866/server/src/main/java/org/opensearch/search/SearchHits.java#L248

owaiskazi19 commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,221 @@
+            String fieldName = parser.currentName();
+            parser.nextToken();
+
+            if (PROCESSOR_NAME_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                processorName = parser.text();
+            } else if (DURATION_MILLIS_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                durationMillis = parser.longValue();
+            } else if (INPUT_DATA_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                inputData = XContentUtils.readValue(parser, parser.currentToken());
+            } else if (OUTPUT_DATA_FIELD.match(fieldName, parser.getDeprecationHandler())) {
+                outputData = XContentUtils.readValue(parser, parser.currentToken());
+            } else {
+                parser.skipChildren();
+            }
```
```suggestion
            switch (fieldName) {
                case PROCESSOR_NAME_FIELD:
                    processorName = parser.text();
                    break;
                case DURATION_MILLIS_FIELD:
                    durationMillis = parser.longValue();
                    break;
                case INPUT_DATA_FIELD:
                    inputData = XContentUtils.readValue(parser, parser.currentToken());
                    break;
                case OUTPUT_DATA_FIELD:
                    outputData = XContentUtils.readValue(parser, parser.currentToken());
                    break;
                default:
                    parser.skipChildren();
                    break;
            }
```

owaiskazi19 commented on server/src/main/java/org/opensearch/search/pipeline/PipelineProcessingContext.java:
```diff
@@ -35,4 +40,33 @@ public void setAttribute(String name, Object value) {
+     */
+    @SuppressWarnings("unchecked")
+    public List<ProcessorExecutionDetail> getProcessorExecutionDetails() {
+        Object details = attributes.get(PROCESSOR_EXECUTION_DETAILS_KEY);
```
Why keeping it as Object here and not directly `List<ProcessorExecutionDetail>` since that's how we added it in the first place? That would reduce the below check

junweid62 replied:
The reason I originally used an Object check was to account for potential flexibility in how values might be stored in attributes. Since attributes is a generic map, I wanted to ensure that even if PROCESSOR_EXECUTION_DETAILS_KEY was accidentally associated with a non-List value, the code would handle it gracefully without throwing a ClassCastException. The instanceof check provided that extra layer of safety.

However, after reviewing the usage of PROCESSOR_EXECUTION_DETAILS_KEY, I confirmed that it is always associated with a List<ProcessorExecutionDetail>, so I simplified the code as suggested.

owaiskazi19 commented on server/src/test/java/org/opensearch/action/search/SearchResponseTests.java:
```diff
@@ -354,8 +361,24 @@ public void testToXContent() {
+                expectedString.append("]");
             }
             expectedString.append("}");
+
```
Remove the extra line

owaiskazi19 reviewed (commented):
Good start.
Few questions
1. Search pipelines can be used 3 ways: using it as default pipeline, as a temporary search request and through query param. Before applying the verbose pipeline param, we should check if search pipeline is applied in the Search Request. You have mentioned that in the RFC in resolvePipeline but I don't see the check in the PR.

2. Once the check is done then need 3 more tests for verbose pipeline execution with all the ways mentioned above

We are touching few core search files. @msfroh can you take a look at it too?

ohltyler commented:
> Tracks each processor’s input, output, execution time, and status (success/failure).
Let's make sure to include the status, plus any associated error message at a per-processor granularity as well.

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -225,17 +247,33 @@ ActionListener<SearchResponse> transformResponseListener(
+                    if (request.source().verbosePipeline()) {
+                        detail.addOutput(Arrays.asList(rr.getHits().deepCopy().getHits()));
+                        detail.addTook(took);
+                        requestContext.addProcessorExecutionDetail(detail);
+                        rr.getInternalResponse().getProcessorResult().add(detail);
```
Do you need to append to both the `requestContext` and to the `SearchResponse`? Adding it to the `requestContext` doesn't help at this point. (It looks like you're done reading from it by the time you get here, since you read from it on line 296, and this code runs bottom-to-top.)

As mentioned in my comment above, I would just wrap each of the processors, let them all write to the `PipelineProcessingContext`, then pass the details from that to the response just before calling `finalListener`.

junweid62 commented:
Hi @msfroh, thank you for your detailed feedback. I have addressed all your comments and made the suggested changes, please take a look and let me know if any further adjustments are needed.

owaiskazi19 reviewed (commented):
Almost there! Good work @junweid62

owaiskazi19 commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,302 @@
+     * @opensearch.internal
+     * @since 2.19.0
+     */
+    @PublicApi(since = "2.19.0")
```
This should be an internal API which is what is mentioned on L293

junweid62 replied:
Thanks for pointing that out! I understand that L293 marks it as @opensearch.internal, but since ProcessorStatus is referenced by public API ProcessorExecutionDetail, it needs to be marked as @PublicApi to comply with OpenSearch's API visibility requirements. I've also updated the comments accordingly. Let me know if you have any other suggestions!

junweid62 commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -156,7 +160,7 @@ void transformRequest(SearchRequest request, ActionListener<SearchRequest> reque
                     afterRequestProcessor(processor, took);
                     onRequestProcessorFailed(processor);
-                    if (processor.isIgnoreFailure()) {
+                    if (processor.isIgnoreFailure() || r.source().verbosePipeline()) {
```
Because when verbosePipeline is enabled, we don't interrupt the search process or immediately return on failure. Instead, we log the failure for the processor and continue executing the rest of the pipeline.

owaiskazi19 commented on server/src/main/java/org/opensearch/search/pipeline/ProcessorExecutionDetail.java:
```diff
@@ -0,0 +1,299 @@
+ *
+ * @opensearch.internal
+ */
+@PublicApi(since = "2.19.0")
```
Why are we keeping this as public? Also, not matching with the comment on L35

junweid62 replied:
This class is intentionally marked as @PublicApi because it is part of the search response and will be exposed to users. I understand the confusion with the comment on L35, and I will revise the comment to make it clearer that this class is indeed a public-facing component. :)

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -156,7 +160,7 @@ void transformRequest(SearchRequest request, ActionListener<SearchRequest> reque
                     afterRequestProcessor(processor, took);
                     onRequestProcessorFailed(processor);
-                    if (processor.isIgnoreFailure()) {
+                    if (processor.isIgnoreFailure() || r.source().verbosePipeline()) {
```
This is essentially saying that the `verbose_pipeline` flag also causes all failures to be ignored. I'm not convinced that's a good idea, since it ends up changing the behavior of the pipeline.

I think you're better off letting failures cascade if `ignoreFailure` is `false`.

junweid62 replied:
Thank you for your feedback! You're absolutely right that this behavior does cause all failures to be ignored. However, those failures are still captured and logged in the ProcessorExecutionDetail, which is exactly what the Flow Framework frontend needs to provide detailed insights. If we were to revert to the original logic, wouldn't users need to explicitly set ignoreFailure themselves when configuring the processor? This additional step might make the setup less intuitive for users who want detailed failure information without impacting the pipeline flow. What do you think?

below are examples of failure processor output 
```
{
            "processor_name": "rename_field",
            "duration_millis": 0,
            "status": "fail",
            "error": "Document with id 1 is missing field messag123e",

```

msfroh replied:
Hmm... I suppose that makes sense. 

In that case, we should clearly document that setting `verbose_pipeline=true`implicitly captures then ignores all processor failures. :+1:

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -235,7 +240,8 @@ ActionListener<SearchResponse> transformResponseListener(
                     long took = TimeUnit.NANOSECONDS.toMillis(relativeTimeSupplier.getAsLong() - start);
                     afterResponseProcessor(processor, took);
-                    if (processor.isIgnoreFailure()) {
+                    // If an error occurs and the pipeline is in verbose mode, the processor will not terminate the execution chain.
+                    if (processor.isIgnoreFailure() || request.source().verbosePipeline()) {
```
Similar comment to above.

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -200,8 +204,8 @@ ActionListener<SearchResponse> transformResponseListener(
         ActionListener<SearchResponse> responseListener,
         PipelineProcessingContext requestContext
     ) {
+        // If there are no response processors, directly check requestContext for execution details
```
Does this comment still apply?

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/PipelinedRequest.java:
```diff
@@ -23,6 +27,7 @@
 public final class PipelinedRequest extends SearchRequest {
     private final Pipeline pipeline;
     private final PipelineProcessingContext requestContext;
+    private static final Logger logger = LogManager.getLogger(Pipeline.class);
```
I think this is a leftover from your previous revision?

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:
```diff
@@ -426,6 +425,11 @@ public PipelinedRequest resolvePipeline(SearchRequest searchRequest, IndexNameEx
                 pipeline = pipelineHolder.pipeline;
             }
         }
+        if (searchRequest.source() != null
+            && searchRequest.source().verbosePipeline()
+            && NOOP_PIPELINE_ID.equals(pipeline.getId()) == true) {
+            throw new IllegalArgumentException("The 'verbose pipeline' option requires a search pipeline to be defined.");
+        }
```
Is this necessary? I guess it doesn't hurt.

Rather than looking at the `NOOP_PIPELINE_ID`, I think you can at least check for object equality against `NO_OP_PIPELINE` (since it's a singleton).

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/TrackingSearchRequestProcessorWrapper.java:
```diff
@@ -0,0 +1,106 @@
+        ProcessorExecutionDetail detail = new ProcessorExecutionDetail(getType(), getTag());
+
+        long start = System.nanoTime();
+        try {
```
Do you need this `try/catch` block? 

In theory, it's the responsibility of the wrapped processor to never throw an exception from the `processRequestAsync` block.

junweid62 replied:
```
} catch (Exception e) {
            detail.markProcessorAsFailed(ProcessorExecutionDetail.ProcessorStatus.FAIL, e.getMessage());
            requestContext.addProcessorExecutionDetail(detail);
```
Thank you for pointing this out! I believe the try/catch block is still necessary in this case because it contains specific logic to handle errors. In particular, the catch block records the error in the ProcessorExecutionDetail, which is essential for tracking failures and providing detailed insights for debugging. This behavior aligns with the requirements of the Flow Framework. Without this block, we would lose the ability to capture and log processor-specific errors.

msfroh replied:
Ahh... but the exceptions will be captured by the async exception listener on line 95. 

If this `catch` block gets hit, it means that an exception was thrown from the call to `processRequestAsync`, which should never be the case. It's the responsibility of `processRequestAsync` implementors to redirect any exceptions to the exception listener.

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/TrackingSearchResponseProcessorWrapper.java:
```diff
@@ -0,0 +1,113 @@
+        ProcessorExecutionDetail detail = new ProcessorExecutionDetail(getType(), getTag());
+
+        long start = System.nanoTime();
+        try {
```
Similar comment to the request processor above. I don't think you need this `try/catch` block.

msfroh reviewed (commented):
Thanks, @junweid62! It's looking pretty good.

I just had a few comments, mostly around code that I think you can safely remove.

owaiskazi19 reviewed (approved):
LGTM! Thanks for patiently addressing all the comments. Can merge after @msfroh review. Please make sure to perform the BWC process on the PR

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/TrackingSearchRequestProcessorWrapper.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public SearchRequest processRequest(SearchRequest request, PipelineProcessingContext requestContext) throws Exception {
+        try {
+            return wrappedProcessor.processRequest(request, requestContext);
+        } catch (UnsupportedOperationException e) {
+            throw e;
+        }
+    }
```
Oh -- we were talking over Slack about this one, but actually I meant that you could throw an `UnsupportedOperationException` from this method.

The only `process*` method that needs to be implemented is `processRequestAsync`. That's the one that gets called from `Pipeline`.

msfroh commented on server/src/main/java/org/opensearch/search/pipeline/TrackingSearchResponseProcessorWrapper.java:
```diff
@@ -0,0 +1,101 @@
+        long start = System.nanoTime();
+        try {
+            detail.addInput(Arrays.asList(response.getHits().deepCopy().getHits()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
If this throws an exception, it must be passed back to the `responseListener` like:

```
} catch (IOException e) {
  responseListener.onFailure(e);
  return;
}
```
